pageserver/src/tenant.rs

```diff
@@ -1461,51 +1622,45 @@ impl Tenant {
-        // We already deleted the layer files, so it's probably best to panic.
-        // (Ideally, above remove_dir_all is atomic so we don't see this timeline after a restart)
-        if children_exist {
-            panic!("Timeline grew children while we removed layer files");
-        }
-        let removed_timeline = timelines.remove(&timeline_id);
-        if removed_timeline.is_none() {
-            // This can legitimately happen if there's a concurrent call to this function.
-            //   T1                                             T2
-            //   lock
-            //   unlock
-            //                                                  lock
-            //                                                  unlock
-            //                                                  remove files
-            //                                                  lock
-            //                                                  remove from map
-            //                                                  unlock
-            //                                                  return
-            //   remove files
-            //   lock
-            //   remove from map observes empty map
-            //   unlock
-            //   return
-            debug!("concurrent call to this function won the race");
+        // we use panicking to poison the tenant if we have grown a new branch while deleting.
+        // FIXME: don't allow branch_timeline while delete_timeline
+        let res = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+            // Remove the timeline from the map.
+            let mut timelines = self.timelines.lock().unwrap();
+            let children_exist = timelines
+                .iter()
+                .any(|(_, entry)| entry.get_ancestor_timeline_id() == Some(timeline.timeline_id));
+            // XXX this can happen because `branch_timeline` doesn't check `TimelineState::Stopping`.
+            // We already deleted the layer files, so it's probably best to panic.
+            // (Ideally, above remove_dir_all is atomic so we don't see this timeline after a restart)
+            if children_exist {
+                // this is to poison the tenants timelines mutex
+                panic!("Timeline grew children while we removed layer files");
+            }
+            let removed_timeline = timelines.remove(&timeline.timeline_id);
+            if removed_timeline.is_none() {
+                debug!("concurrent call to this function won the race");
+            }
+            drop(timelines);
+        }));
+
+        match res {
+            Ok(()) => Ok(()),
+            Err(_) => {
+                // not going to continue unwind, just fail.
+                // the panick has been already logged and reported, hopefully nothing got to start
+                // with this new timeline.
+                Err(InnerDeleteTimelineError::ChildAppearedAfterRemoveDir)
```

Comment:
Should explain here what happens on retry.
My understanding is we'll try to lock the ()`self`) timeline lock, likely unwrap, and panic?

Also, suggestion on the comment to avoid confusion which timeline is being referred to:

```suggestion
                // not going to continue unwind, just fail.
                // the panick has been already logged and reported, hopefully nothing got to start
                // with the child timeline.
                Err(InnerDeleteTimelineError::ChildAppearedAfterRemoveDir)
```

Reply:
This is a permanent error, should not be retried. Yes in general with a new callsite, if we would get to starting over and **not panic at the HasChildren checks**, we would get a poisonerror unwrap within this catch_unwind.

Unsure if this needs to be explained, because this should not be retried.

Reply:
In #4194 this is now `DeletedGrewChildren`, continues to be terminal state.